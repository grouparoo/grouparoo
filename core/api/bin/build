#!/bin/bash

set -e

cd "$(dirname "$0")"
cd ..

echo -e "Building @grouparoo/core API..."

START_TIME=$(date +%s);

rm -rf dist
./../node_modules/.bin/tsc --declaration

END_TIME=$(date +%s);

echo -e "@grouparoo/core API built in $(($END_TIME - $START_TIME))s"
echo -e ""
