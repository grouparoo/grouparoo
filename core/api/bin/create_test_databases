#!/usr/bin/env bash

N=$(($(getconf _NPROCESSORS_ONLN))) || 8

if [ $1 ]; then
  echo "creating database grouparoo_test_$1"
  if [ -n "$CI" ]; then
    `PGPASSWORD=postgres createdb "grouparoo_test_$1" --host postgres --user postgres`
  else
    `createdb "grouparoo_test_$1"`
  fi
else
  for ((i=1; i<=$N; i++)) {
    echo "creating database grouparoo_test_$i"
    if [ -n "$CI" ]; then
      `PGPASSWORD=postgres createdb "grouparoo_test_$i" --host postgres --user postgres`
    else
      `createdb "grouparoo_test_$i"`
    fi
  }
fi


