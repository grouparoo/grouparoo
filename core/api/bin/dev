#!/usr/bin/env bash

cd "$(dirname "$0")"
cd ..

if [ -f "../node_modules/.bin/ts-node-dev" ]; then
  EXECUTABLE="../node_modules/.bin/ts-node-dev"
else
  EXECUTABLE="../../../../node_modules/.bin/ts-node-dev"
fi

export NEXT_TELEMETRY_DISABLED=true
export NEXT_DEVELOPMENT_MODE=true

# in develeopment mode, often there is an orphan next.js/webpack process hanging around if the process crashes.
GROUPAROO_PIDS=`ps | grep ts-node-dev | grep grouparoo | cut -f1 -d' '`
for p in $GROUPAROO_PIDS; do
  printf "stopping orphan process $p\r\n"
  kill -9 $p
done


node \
  --max-old-space-size=4096 \
  -- "$EXECUTABLE" \
    --transpile-only \
    --ignore-watch=web \
    --ignore-watch='../web' \
    --no-deps \
    --notify=false \
      -- \
      src/server.ts
