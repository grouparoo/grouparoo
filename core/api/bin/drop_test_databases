#!/usr/bin/env bash

N=$(($(getconf _NPROCESSORS_ONLN))) || 8

if [ $1 ]; then
  echo "dropping database grouparoo_test_$1"
  if [ -n "$CI" ]; then
    `PGPASSWORD=postgres dropdb "grouparoo_test_$1" --host postgres --user postgres`
  else
    `dropdb "grouparoo_test_$1"`
  fi
else
  for ((i=1; i<=$N; i++)) {
    echo "dropping database grouparoo_test_$i"
    if [ -n "$CI" ]; then
      `PGPASSWORD=postgres dropdb "grouparoo_test_$i" --host postgres --user postgres`
    else
      `dropdb "grouparoo_test_$i"`
    fi
  }
fi
