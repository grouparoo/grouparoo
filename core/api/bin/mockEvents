#!/usr/bin/env node

const { Process, log, api, config } = require("actionhero");
const { ApiKey, Profile } = require("../dist");
const optimist = require("optimist");
const fetch = require("isomorphic-fetch");
const uuid = require("uuid");

// prettier-ignore
const funnel = [
  { type: "pageview", data: { page: "/" }, percent: 100 },
  { type: "pageview", data: { page: "/about" }, percent: 90 },
  { type: "pageview", data: { page: "/sign-in" }, percent: 90 },
  { type: "identify", data: { page: "/sign-in" }, percent: 100 },
  { type: "pageview", data: { page: '/product'}, percent: 80, addToCartPercent : 50 },
  { type: "pageview", data: { page: '/product'}, percent: 80, addToCartPercent : 50 },
  { type: "pageview", data: { page: '/product'},  percent: 80, addToCartPercent : 50 },
  { type: "pageview", data: { page: '/cart'},  percent: 75 },
  { type: "pageview", data: { page: '/purchase'},  percent: 80 },
  { type: "pageview", data: { page: '/thanks'}, percent: 100 },
];

const products = [
  { name: "red-shell", price: 100 },
  { name: "blue-shell", price: 150 },
  { name: "green-shell", price: 50 },
  { name: "coin", price: 1 },
  { name: "banana-peel", price: 25 },
  { name: "mushroom", price: 40 },
  { name: "star", price: 99 },
];

const sleep = 100;
const parallelSessions = 1;

async function main() {
  const app = new Process();
  await app.initialize();

  if (!optimist.argv["apiKey"]) {
    throw new Error(`argument apiKey is required`);
  }
  const { apiKey } = await ApiKey.findOne({
    where: { apiKey: optimist.argv["apiKey"] },
  });
  if (!apiKey) {
    throw new Error("apiKey not found (--apiKey=abc123)");
  }

  if (!optimist.argv["url"]) {
    throw new Error(`argument url is required (--url="http://localhost:3000")`);
  }
  const baseUrl = `${optimist.argv["url"]}/api/v1/track`;

  const sessions = [];
  for (let i = 0; i < parallelSessions; i++) {
    sessions.push(new MockSession(i, baseUrl, apiKey));
  }

  async function tick() {
    for (const i in sessions) {
      await sessions[i].tick();
    }

    setTimeout(tick, sleep);
  }

  await tick();
}

class MockSession {
  constructor(id, baseUrl, apiKey) {
    this.id = id;
    this.baseUrl = baseUrl;
    this.apiKey = apiKey;
    this.anonymousId = null;
    this.currentUser = null;
    this.profileProperties = {};
    this.currentStep = 0;
    this.identified = false;
    this.cart = [];
  }

  log(message) {
    log(`[session ${this.id}] - ${message}`);
  }

  async tick() {
    if (!this.currentStep) {
      this.currentStep = 0;
    }
    const step = funnel[this.currentStep];
    await this.setCurrentUser();
    await this.checkPercent(step);
    if (!this.currentUser) {
      return;
    }

    await this.track(step);

    this.currentStep++;
    if (this.currentStep >= funnel.length) {
      await this.empty();
    }
  }

  async track(step) {
    const params = {
      apiKey: this.apiKey,
      anonymousId: this.anonymousId,
      occurredAt: new Date().getTime(),
      type: step.type,
      data: step.data || {},
    };

    if (step.type === "identify") {
      this.identified = true;
    }
    if (this.identified) {
      params.userId = this.profileProperties["userId"].value;
    }

    let relatedEventParams = {
      apiKey: this.apiKey,
      anonymousId: this.anonymousId,
      occurredAt: new Date().getTime(),
      userId: this.profileProperties["userId"].value,
    };

    if (step.type === "pageview" && step.data.page === "/product") {
      const currentProduct =
        products[Math.floor(Math.random() * products.length)];
      const randomNumber = Math.floor(Math.random() * 101);
      const added = randomNumber <= step.addToCartPercent;
      if (added) {
        this.cart.push(currentProduct);
        relatedEventParams.type = "itemAddedToCart";
        relatedEventParams.data = {
          name: currentProduct.name,
          price: currentProduct.price,
        };
      }
    } else if (step.type === "pageview" && step.data.page === "/purchase") {
      relatedEventParams.type = "purchase";
      let total = 0;
      this.cart.map((product) => (total = total + product.price));
      relatedEventParams.data = {
        products: JSON.stringify(this.cart.map((product) => product.name)),
        total,
      };
    }

    const response = await fetch(this.baseUrl, {
      method: "post",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(params),
    });
    this.log(this.baseUrl + " " + JSON.stringify(params));

    if (relatedEventParams.type) {
      const relatedEventResponse = await fetch(this.baseUrl, {
        method: "post",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify(relatedEventParams),
      });
      this.log(this.baseUrl + " " + JSON.stringify(relatedEventParams));
    }
  }

  async setCurrentUser() {
    if (this.currentUser) {
      return;
    }

    this.currentUser = await Profile.findOne({
      order: api.sequelize.random(),
    });
    this.anonymousId = `?-${uuid.v4()}`;
    this.profileProperties = await this.currentUser.properties();
    if (!this.profileProperties["userId"]) {
      return this.empty();
    }
    this.log(
      `using profile ${this.currentUser.guid} (${this.profileProperties["userId"].value})`
    );
  }

  async checkPercent(step) {
    const randomNumber = Math.floor(Math.random() * 101);
    if (randomNumber > step.percent) {
      return this.empty();
    }
  }

  async empty() {
    this.currentUser = null;
    this.profileProperties = {};
    this.currentStep = 0;
    this.identified = false;
    this.cart = [];
    this.log(`session complete`);
  }
}

(async function () {
  await main();
})();
