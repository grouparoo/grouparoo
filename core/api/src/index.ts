export { plugin } from "./modules/plugin";

export * from "./classes/plugin";
export * from "./classes/fileTransport";

export { App } from "./models/App";
export { ApiKey } from "./models/ApiKey";
export { Destination } from "./models/Destination";
export { DestinationGroup } from "./models/DestinationGroup";
export { DestinationGroupMembership } from "./models/DestinationGroupMembership";
export { Event } from "./models/Event";
export { EventData } from "./models/EventData";
export { Export } from "./models/Export";
export { File } from "./models/File";
export { Group } from "./models/Group";
export { GroupMember } from "./models/GroupMember";
export { GroupRule } from "./models/GroupRule";
export { Import } from "./models/Import";
export { Log } from "./models/Log";
export { Mapping } from "./models/Mapping";
export { Option } from "./models/Option";
export { Permission } from "./models/Permission";
export { Profile } from "./models/Profile";
export { ProfileProperty } from "./models/ProfileProperty";
export { ProfilePropertyRule } from "./models/ProfilePropertyRule";
export { ProfilePropertyRuleFilter } from "./models/ProfilePropertyRuleFilter";
export { Run } from "./models/Run";
export { Schedule } from "./models/Schedule";
export { Setting } from "./models/Setting";
export { Source } from "./models/Source";
export { Team } from "./models/Team";
export { TeamMember } from "./models/TeamMember";

export { waitForLock } from "./modules/locks";
