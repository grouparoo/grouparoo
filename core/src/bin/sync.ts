import { GrouparooCLI } from "../modules/cli";
import { GrouparooRecord, RecordProperty, Property } from "..";
import { CLI, ParamsFrom } from "actionhero";
import { Op } from "sequelize";

export class SyncCLI extends CLI {
  name = "sync <recordProperty>";
  description =
    "Sync (import & export) a GrouparooRecord.  You can provide a GrouparooRecord id or a unique GrouparooRecord Property";
  inputs = {
    property: {
      description:
        "Choose the name of the Property to find the GrouparooRecord by (i.e.: email_address)",
      letter: "p",
    },
    "no-export": {
      description: "Skip exporting the record",
      letter: "n",
      flag: true,
    },
  };
  example = `grouparoo sync person@example.com --property email`;

  constructor() {
    super();
    GrouparooCLI.timestampOption(this);
    GrouparooCLI.jsonOption(this);
  }

  preInitialize = () => {
    GrouparooCLI.setGrouparooRunMode(this);
    GrouparooCLI.setNextDevelopmentMode();
  };

  async run({
    params,
  }: {
    params: Partial<
      ParamsFrom<SyncCLI> & {
        export?: boolean;
        recordProperty?: string;
        json?: string;
        _arguments: string[];
      }
    >;
  }) {
    const [recordProperty] = params._arguments || [];
    if (recordProperty) params.recordProperty = recordProperty;

    if (!params.recordProperty) {
      return GrouparooCLI.logger.fatal("No GrouparooRecord Property provided");
    }

    if (typeof params.property === "boolean") {
      return GrouparooCLI.logger.fatal("Provide the property's name");
    }

    if (!params.json) {
      GrouparooCLI.logCLI(
        this.name.replace(
          " <recordProperty>",
          params.recordProperty ? " " + params.recordProperty : ""
        )
      );
    }

    const uniqueProperties = (
      await Property.findAll({ where: { unique: true } })
    ).filter((p) => (params.property ? p.key === params.property : true));

    const recordProperties = await RecordProperty.findAll({
      where: {
        rawValue: params.recordProperty,
        propertyId: { [Op.in]: uniqueProperties.map((p) => p.id) },
      },
    });

    const wheres = [];
    if (recordProperties.length > 1) {
      return GrouparooCLI.logger.fatal(
        `More than one GrouparooRecord found with GrouparooRecord Property "${params.recordProperty}".  Please provide [property]`
      );
    } else if (recordProperties.length === 1) {
      wheres.push({ id: recordProperties[0].recordId });
    } else {
      wheres.push({ id: params.recordProperty });
    }

    const record = await GrouparooRecord.findOne({
      where: { [Op.or]: wheres },
    });

    if (!record) {
      return GrouparooCLI.logger.fatal(
        params.property
          ? `Cannot find GrouparooRecord where GrouparooRecord Property "${params.property}"="${params.recordProperty}"`
          : `Cannot find GrouparooRecord with id or unique GrouparooRecord Property "${params.recordProperty}"`
      );
    }

    const { properties, groups, exports } = await record.snapshot(
      params.export
    );

    if (params.json) {
      GrouparooCLI.logger.log(JSON.stringify({ properties, groups, exports }));
    } else {
      const propertyStatus: Record<string, any> = {};
      const groupStatus: Record<string, any> = {};
      const exportStatus: Record<string, any> = {};
      for (const k in properties)
        propertyStatus[k] = [properties[k].values.join(", ")];
      for (const i in groups) groupStatus[groups[i].name] = [];
      for (const i in exports) {
        exportStatus[exports[i].destinationName + ": Properties"] = [
          Object.keys(exports[i].newRecordProperties)
            .map((k) => `${k}: ${exports[i].newRecordProperties[k]}`)
            .join(", "),
        ];
        exportStatus[exports[i].destinationName + ": Groups"] = [
          exports[i].newGroups.join(", "),
        ];
      }

      GrouparooCLI.logger.status(
        `GrouparooRecord`,
        [
          {
            header: "Info",
            status: {
              id: [record.id],
              createdAt: [record.createdAt.toISOString()],
              state: [record.state],
            },
          },
          { header: "GrouparooRecord Properties", status: propertyStatus },
          { header: "Groups", status: groupStatus },
          {
            header: params.export ? "Exports" : "Next Exports",
            status: exportStatus,
          },
        ],
        record.id
      );
    }

    return true;
  }
}
