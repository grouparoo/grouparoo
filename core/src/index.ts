export { plugin } from "./modules/plugin";

export * from "./classes/plugin";
export * from "./classes/fileTransport";
export * from "./classes/configTemplate";
export { AuthenticatedAction } from "./classes/actions/authenticatedAction";
export { OptionallyAuthenticatedAction } from "./classes/actions/optionallyAuthenticatedAction";

export { App, SimpleAppOptions } from "./models/App";
export { ApiKey } from "./models/ApiKey";
export {
  Destination,
  SimpleDestinationOptions,
  DestinationSyncMode,
  DestinationSyncOperations,
} from "./models/Destination";
export { DestinationGroupMembership } from "./models/DestinationGroupMembership";
export { Export } from "./models/Export";
export { ExportProcessor } from "./models/ExportProcessor";
export { File } from "./models/File";
export { Group } from "./models/Group";
export { GroupMember } from "./models/GroupMember";
export { GroupRule } from "./models/GroupRule";
export { Import } from "./models/Import";
export { Log } from "./models/Log";
export { Notification } from "./models/Notification";
export { Mapping } from "./models/Mapping";
export { Option } from "./models/Option";
export { Permission } from "./models/Permission";
export { GrouparooRecord } from "./models/GrouparooRecord";
export { RecordProperty } from "./models/RecordProperty";
export {
  Property,
  PropertyTypes,
  PropertyFiltersWithKey,
  SimplePropertyOptions,
} from "./models/Property";
export { Filter } from "./models/Filter";
export { RecordMultipleAssociationShim } from "./models/RecordMultipleAssociationShim";
export { Run, HighWaterMark } from "./models/Run";
export {
  Schedule,
  SimpleScheduleOptions,
  PluginConnectionScheduleOption,
  ScheduleFiltersWithKey,
} from "./models/Schedule";
export { Session } from "./models/Session";
export { Setting } from "./models/Setting";
export { SetupStep } from "./models/SetupStep";
export { Source, SimpleSourceOptions, SourceMapping } from "./models/Source";
export { Team } from "./models/Team";
export { TeamMember } from "./models/TeamMember";

export { Status } from "./modules/status";
export { StatusMetric } from "./modules/statusReporters";
export { GrouparooCLI } from "./modules/cli";

export { waitForLock } from "./modules/locks";
export { Errors } from "./modules/errors";
export * from "./modules/cache";
export * from "./modules/tableSpeculation";
