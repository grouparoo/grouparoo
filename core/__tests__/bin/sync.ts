import { helper, relaxedSnapshot } from "@grouparoo/spec-helper";
import { SyncCLI } from "../../src/bin/sync";
import { api } from "actionhero";
import { GrouparooPlugin, GrouparooRecord } from "../../src";

describe("bin/sync", () => {
  helper.grouparooTestServer({ truncate: true, enableTestPlugin: true });
  beforeAll(async () => await helper.factories.properties());

  let messages: string[] = [];
  const spies: jest.SpyInstance[] = [];
  let record: GrouparooRecord;

  beforeAll(async () => {
    // we need to ensure that we have static results for the import to snapshot
    const testPlugin: GrouparooPlugin = api.plugins.plugins.find(
      (a) => a.name === "@grouparoo/test-plugin"
    );
    const testPluginConnection = testPlugin.connections.find(
      (c) => c.name === "test-plugin-import"
    );
    testPluginConnection.methods.recordProperty = async ({ property }) => {
      const data = {
        userId: [100],
        isVIP: [true],
        email: [`mario@example.com`],
        firstName: ["Mario"],
        lastName: ["Mario"],
        ltv: [100.0],
        lastLoginAt: [new Date(0)],
        purchases: ["hat", "mushroom"],
        purchaseAmounts: [100, 200],
      };

      return data[property.key as keyof typeof data];
    };
  });

  beforeAll(async () => (record = await helper.factories.record()));

  beforeEach(() => {
    messages = [];
    spies.push(
      jest
        .spyOn(console, "log")
        .mockImplementation((message) => messages.push(message))
    );
    spies.push(
      jest
        .spyOn(console, "error")
        .mockImplementation((message) => messages.push(message))
    );
  });

  afterEach(() => {
    spies.map((s) => s.mockRestore());
  });

  describe("with instance", () => {
    let instance: SyncCLI;
    beforeAll(() => {
      instance = new SyncCLI();
    });

    test("it will fail without a record property", async () => {
      await instance.run({ params: {} });
      const output = messages.join(" ");
      expect(output).toContain("No GrouparooRecord Property provided");
    });

    test("it can sync by record id", async () => {
      await instance.run({ params: { recordProperty: record.id } });
      const output = messages.join(" ");
      expect(output).toContain(`* id: ${record.id}`);
      expect(output).toContain(`* email: mario@example.com`);
    });

    test("it can sync by record property", async () => {
      await instance.run({ params: { recordProperty: "mario@example.com" } });
      const output = messages.join(" ");
      expect(output).toContain(`* id: ${record.id}`);
      expect(output).toContain(`* email: mario@example.com`);
    });

    test("it can sync by record property and property", async () => {
      await instance.run({
        params: { recordProperty: "mario@example.com", property: "email" },
      });
      const output = messages.join(" ");
      expect(output).toContain(`* id: ${record.id}`);
      expect(output).toContain(`* email: mario@example.com`);
    });

    test("it will fail with if it cannot provide the explicit property", async () => {
      await instance.run({
        params: { recordProperty: "mario@example.com", property: "foo" },
      });
      const output = messages.join(" ");
      expect(output).toContain(
        'Cannot find GrouparooRecord where GrouparooRecord Property "foo"="mario@example.com"'
      );
    });

    test("it can return JSON", async () => {
      process.env.GROUPAROO_LOG_LEVEL = "alert"; // disable other log messages

      await instance.run({
        params: { recordProperty: "mario@example.com", json: "true" },
      });
      const output = messages.join("");

      const json = JSON.parse(output);

      expect(json).toMatchSnapshot(relaxedSnapshot(json)); // snapshot test
      expect(json.properties.userId.values).toEqual([100]); // direct expectation
    });
  });
});
