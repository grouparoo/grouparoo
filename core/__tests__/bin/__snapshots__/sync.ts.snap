// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bin/sync with instance it can return JSON 1`] = `
Object {
  "exports": Array [],
  "groups": Array [],
  "properties": Object {
    "email": Object {
      "confirmedAt": Any<String>,
      "createdAt": Any<String>,
      "directlyMapped": false,
      "id": Any<String>,
      "identifying": false,
      "isArray": false,
      "startedAt": null,
      "state": "ready",
      "stateChangedAt": Any<String>,
      "type": "email",
      "unique": true,
      "updatedAt": Any<String>,
      "valueChangedAt": Any<String>,
      "values": Array [
        "mario@example.com",
      ],
    },
    "firstName": Object {
      "confirmedAt": Any<String>,
      "createdAt": Any<String>,
      "directlyMapped": false,
      "id": Any<String>,
      "identifying": false,
      "isArray": false,
      "startedAt": null,
      "state": "ready",
      "stateChangedAt": Any<String>,
      "type": "string",
      "unique": false,
      "updatedAt": Any<String>,
      "valueChangedAt": Any<String>,
      "values": Array [
        "Mario",
      ],
    },
    "isVIP": Object {
      "confirmedAt": Any<String>,
      "createdAt": Any<String>,
      "directlyMapped": false,
      "id": Any<String>,
      "identifying": false,
      "isArray": false,
      "startedAt": null,
      "state": "ready",
      "stateChangedAt": Any<String>,
      "type": "boolean",
      "unique": false,
      "updatedAt": Any<String>,
      "valueChangedAt": Any<String>,
      "values": Array [
        true,
      ],
    },
    "lastLoginAt": Object {
      "confirmedAt": Any<String>,
      "createdAt": Any<String>,
      "directlyMapped": false,
      "id": Any<String>,
      "identifying": false,
      "isArray": false,
      "startedAt": null,
      "state": "ready",
      "stateChangedAt": Any<String>,
      "type": "date",
      "unique": false,
      "updatedAt": Any<String>,
      "valueChangedAt": Any<String>,
      "values": Array [
        "1970-01-01T00:00:00.000Z",
      ],
    },
    "lastName": Object {
      "confirmedAt": Any<String>,
      "createdAt": Any<String>,
      "directlyMapped": false,
      "id": Any<String>,
      "identifying": false,
      "isArray": false,
      "startedAt": null,
      "state": "ready",
      "stateChangedAt": Any<String>,
      "type": "string",
      "unique": false,
      "updatedAt": Any<String>,
      "valueChangedAt": Any<String>,
      "values": Array [
        "Mario",
      ],
    },
    "ltv": Object {
      "confirmedAt": Any<String>,
      "createdAt": Any<String>,
      "directlyMapped": false,
      "id": Any<String>,
      "identifying": false,
      "isArray": false,
      "startedAt": null,
      "state": "ready",
      "stateChangedAt": Any<String>,
      "type": "float",
      "unique": false,
      "updatedAt": Any<String>,
      "valueChangedAt": Any<String>,
      "values": Array [
        100,
      ],
    },
    "purchaseAmounts": Object {
      "confirmedAt": Any<String>,
      "createdAt": Any<String>,
      "directlyMapped": false,
      "id": Any<String>,
      "identifying": false,
      "isArray": true,
      "startedAt": null,
      "state": "ready",
      "stateChangedAt": Any<String>,
      "type": "float",
      "unique": false,
      "updatedAt": Any<String>,
      "valueChangedAt": Any<String>,
      "values": Array [
        100,
        200,
      ],
    },
    "purchases": Object {
      "confirmedAt": Any<String>,
      "createdAt": Any<String>,
      "directlyMapped": false,
      "id": Any<String>,
      "identifying": false,
      "isArray": true,
      "startedAt": null,
      "state": "ready",
      "stateChangedAt": Any<String>,
      "type": "string",
      "unique": false,
      "updatedAt": Any<String>,
      "valueChangedAt": Any<String>,
      "values": Array [
        "hat",
        "mushroom",
      ],
    },
    "userId": Object {
      "confirmedAt": Any<String>,
      "createdAt": Any<String>,
      "directlyMapped": true,
      "id": Any<String>,
      "identifying": true,
      "isArray": false,
      "startedAt": null,
      "state": "ready",
      "stateChangedAt": Any<String>,
      "type": "integer",
      "unique": true,
      "updatedAt": Any<String>,
      "valueChangedAt": Any<String>,
      "values": Array [
        100,
      ],
    },
  },
}
`;
